mesos-client/src/main/scala/com/mesosphere/mesos/client/Session.scala

```diff
@@ -0,0 +1,171 @@
+    */
+  private def connection(implicit system: ActorSystem, mat: Materializer): Flow[HttpRequest, HttpResponse, NotUsed] = {
+    val poolSettings = ConnectionPoolSettings("").withMaxConnections(1).withPipeliningLimit(1)
+    if (isSecured) {
+      Flow[HttpRequest]
+        .map(_ -> NotUsed)
+        .via(Http().newHostConnectionPoolHttps(host = url.getHost, port = port, settings = poolSettings))
+        .map {
+          case (Success(response), NotUsed) => response
+          case (Failure(ex), NotUsed) => throw ex
+        }
+    } else {
+      Flow[HttpRequest]
+        .map(_ -> NotUsed)
+        .via(Http().newHostConnectionPool(host = url.getHost, port = port, settings = poolSettings))
+        .map {
+          case (Success(response), NotUsed) => response
+          case (Failure(ex), NotUsed) => throw ex
+        }
+    }
+  }
+}
+
+object SessionActor {
+  def props(
+      credentialsProvider: CredentialsProvider,
+      requestFactory: (Array[Byte], Option[HttpCredentials]) => HttpRequest): Props = {
+    Props(new SessionActor(credentialsProvider, requestFactory))
+  }
+
+  /**
+    * A simple class that captures the original sender of a call and the serialized call.
+    *
+    * @param originalCall The serialized Mesos call.
+    * @param originalSender The sender from the stream.
+    * @param response The response for the call.
+    */
+  private[client] case class Response(originalCall: Array[Byte], originalSender: ActorRef, response: HttpResponse)
+}
+
+/**
+  * An actor makes requests to Mesos using credentials from a [[CredentialsProvider]].
+  *
+  * @param credentialsProvider The provider used to fetch the credentials, ie a session token.
+  * @param requestFactory The factory method that creates a [[HttpRequest]] from a serialized Mesos call. It should
+  *                       capture the [[Session.streamId]] and [[Session.url]].
+  */
+class SessionActor(
+    credentialsProvider: CredentialsProvider,
+    requestFactory: (Array[Byte], Option[HttpCredentials]) => HttpRequest)
+    extends Actor
+    with Stash
+    with StrictLogging {
+
+  import context.dispatcher
+
+  import akka.pattern.pipe
+
+  override def preStart(): Unit = {
+    super.preStart()
+    credentialsProvider.nextToken().pipeTo(self)
+  }
+
+  override def receive: Receive = initializing
+
+  def initializing: Receive = {
+    case credentials: HttpCredentials =>
+      logger.info("Retrieved token")
+      context.become(initialized(credentials))
+      unstashAll()
+    case Status.Failure(ex) =>
+      logger.error("Fetching the next token failed", ex)
+      throw ex
+    case _ => stash()
+  }
+
+  def initialized(credentials: HttpCredentials): Receive = {
+    case call: Array[Byte] =>
+      val request = requestFactory(call, Some(credentials))
+      val originalSender = sender()
+      logger.info("Processing next call.")
+      // The TLS handshake for each connection might be an overhead. We could potentially reuse a connection.
```

Comment:
@timcharper and @meichstedt, I will not implement connection reuse in this PR. I cut scope to get the Keep Alive Framework and the new Jenkins on Soak and MWT. We can always redesign the `SessionActor` to reuse TLS sessions.